src/coreclr/jit/morph.cpp

```diff
@@ -275,15 +294,22 @@ GenTree* Compiler::fgMorphExpandCast(GenTreeCast* tree)
     }
 #endif //! TARGET_64BIT
 
-#ifdef TARGET_ARMARCH
+#if defined(TARGET_ARMARCH) || defined(TARGET_AMD64)
     // AArch, unlike x86/amd64, has instructions that can cast directly from
     // all integers (except for longs on AArch32 of course) to floats.
+    //
+    // AMD64, however, has efficient alternatives that can directly handle
+    // the cases that it doesn't support (uint and ulong). For u64->f32
+    // we want to keep the behavior as `u64->f64->f32` however, as its simpler
+    //
     // Because there is no IL instruction conv.r4.un, uint/ulong -> float
     // casts are always imported as CAST(float <- CAST(double <- uint/ulong)).
     // We can eliminate the redundant intermediate cast as an optimization.
     else if ((dstType == TYP_FLOAT) && (srcType == TYP_DOUBLE) && oper->OperIs(GT_CAST)
-#ifdef TARGET_ARM
-             && !varTypeIsLong(oper->AsCast()->CastOp())
+#if defined(TARGET_ARM)
+             && !varTypeIsLong(oper->AsCast()->CastFromType())
+#elif defined(TARGET_AMD64)
+             && (!varTypeIsLong(oper->AsCast()->CastFromType()) || !oper->AsCast()->IsUnsigned())
```

Comment:
This is a minor optimization. I would think it is better to leave it out of this change and do separately (such that the diffs are also separate, etc).

Reply:
It's not so minor and is actually part of maintaining perf. Keeping it in the same PR will help more easily track what the actual regression here is going to be.